libshaderc/src/shaderc_cpp_test.cc

```diff
@@ -499,12 +584,102 @@ TEST_F(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
   options_.SetDisassemblyMode();
   options_.SetPreprocessingOnlyMode();
   shaderc::SpvModule result_disassembly_mode_first = compiler_.CompileGlslToSpv(
-      kMinimalShaderWithMacro, shaderc_glsl_vertex_shader, options_);
+      kMinimalShaderWithMacro, shaderc_glsl_forced_vertex_shader, options_);
   EXPECT_TRUE(result_disassembly_mode_first.GetSuccess());
   EXPECT_THAT(result_disassembly_mode_first.GetData(),
               HasSubstr("void main(){ }"));
 }
 
+// A shader kind test cases needs: 1) A shader text with or without #pragma
+// annotation, 2) forced_shader_kind, 3) weak_shader_kind.
+struct ShaderKindTestCase {
+  const char* shader_;
+  shaderc_shader_kind shader_kind_;
+};
+
+// Test the shader kind deduction process. If the shader kind is one of the
+// forced ones, the compiler will just try to compile the source code in that
+// specified shader kind. If the shader kind is shaderc_glsl_deduce_from_pragma,
+// the compiler will determine the shader kind from #pragma annotation in the
+// source code and emit error if none such annotation is found. When the shader
+// kind is one of the default ones, the compiler will fall back to use the
+// specified shader kind if and only if #pragma annoation is not found.
+class ShaderKindDeductionTests
+    : public testing::TestWithParam<ShaderKindTestCase> {
+ protected:
+  bool CompilesToValidSpv() {
+    ShaderKindTestCase test_case = GetParam();
+    return IsValidSpv(compiler_.CompileGlslToSpv(
+        test_case.shader_, test_case.shader_kind_, options_));
+  }
+
+  shaderc::Compiler compiler_;
+  CompileOptions options_;
+
+ private:
+  static bool IsValidSpv(const shaderc::SpvModule& result) {
```

Comment:
Maybe this should just be a free function in the module.   It's not really coupled to the internals of the class.

Also, the method needs a blurb.